numpy/lib/nanfunctions.py

```diff
@@ -106,6 +109,8 @@ def _replace_nan(a, val):
         mask = None
 
     if mask is not None:
+        if not input_copied:
```

Comment:
This if statement seems like a premature optimisation to me. Speeding up cases when the input isn't even an array probably isn't a good use of code complexity.

I'd do the copy below unconditionally, and use `asanyarray` on the first line.

Reply:
Fair enough - a second array creation is just noise compared to list iteration. Done.